README
NAME
    App::cpanminus - get, unpack, build and install modules from CPAN

SYNOPSIS
        cpanm Module

    Run "cpanm -h" or "perldoc cpanm" for more options.

DESCRIPTION
    cpanminus is a script to get, unpack, build and install modules from
    CPAN and does nothing else.

    It's dependency free (can bootstrap itself), requires zero
    configuration, and stands alone. When running, it requires only 10MB of
    RAM.

INSTALLATION
    There are several ways to install cpanminus to your system.

  Package management system
    There are Debian packages, RPMs, FreeBSD ports, and packages for other
    operation systems available. If you want to use the package management
    system, search for cpanminus and use the appropriate command to install.
    This makes it easy to install "cpanm" to your system without thinking
    about where to install, and later upgrade.

  Installing to system perl
    You can also use the latest cpanminus to install cpanminus itself:

        curl -L https://cpanmin.us | perl - --sudo App::cpanminus

    This will install "cpanm" to your bin directory like "/usr/local/bin"
    and you'll need the "--sudo" option to write to the directory, unless
    you configured "INSTALL_BASE" with local::lib.

  Installing to local perl (perlbrew)
    If you have perl in your home directory, which is the case if you use
    tools like perlbrew, you don't need the "--sudo" option, since you're
    most likely to have a write permission to the perl's library path. You
    can just do:

        curl -L https://cpanmin.us | perl - App::cpanminus

    to install the "cpanm" executable to the perl's bin path, like
    "~/perl5/perlbrew/bin/cpanm".

  Downloading the standalone executable
    You can also copy the standalone executable to whatever location you'd
    like.

        cd ~/bin
        curl -LO http://xrl.us/cpanm
        chmod +x cpanm
        # edit shebang if you don't have /usr/bin/env

    This just works, but be sure to grab the new version manually when you
    upgrade because "--self-upgrade" might not work for this.

DEPENDENCIES
    perl 5.8 or later.

    *   'tar' executable (bsdtar or GNU tar version 1.22 are recommended) or
        Archive::Tar to unpack files.

    *   C compiler, if you want to build XS modules.

    *   make

    *   Module::Build (core in 5.10)

QUESTIONS
  Another CPAN installer?
    OK, the first motivation was this: the CPAN shell runs out of memory (or
    swaps heavily and gets really slow) on Slicehost/linode's most
    affordable plan with only 256MB RAM. Should I pay more to install perl
    modules from CPAN? I don't think so.

  But why a new client?
    First of all, let me be clear that CPAN and CPANPLUS are great tools
    I've used for *literally* years (you know how many modules I have on
    CPAN, right?). I really respect their efforts of maintaining the most
    important tools in the CPAN toolchain ecosystem.

    However, for less experienced users (mostly from outside the Perl
    community), or even really experienced Perl developers who know how to
    shoot themselves in their feet, setting up the CPAN toolchain often
    feels like yak shaving, especially when all they want to do is just
    install some modules and start writing code.

  Zero-conf? How does this module get/parse/update the CPAN index?
    It queries the CPAN Meta DB site at <http://cpanmetadb.plackperl.org/>.
    The site is updated at least every hour to reflect the latest changes
    from fast syncing mirrors. The script then also falls back to query the
    module at <http://metacpan.org/> using its wonderful API.

    Upon calling these API hosts, cpanm (1.6004 or later) will send the
    local perl versions to the server in User-Agent string by default. You
    can turn it off with "--no-report-perl-version" option. Read more about
    the option with cpanm, and read more about the privacy policy about this
    data collection at <http://cpanmetadb.plackperl.org/#privacy>

    Fetched files are unpacked in "~/.cpanm" and automatically cleaned up
    periodically. You can configure the location of this with the
    "PERL_CPANM_HOME" environment variable.

  Where does this install modules to? Do I need root access?
    It installs to wherever ExtUtils::MakeMaker and Module::Build are
    configured to (via "PERL_MM_OPT" and "PERL_MB_OPT"). So if you're using
    local::lib, then it installs to your local perl5 directory. Otherwise it
    installs to the site_perl directory that belongs to your perl.

    cpanminus at a boot time checks whether you have configured local::lib,
    or have the permission to install modules to the site_perl directory. If
    neither, it automatically sets up local::lib compatible installation
    path in a "perl5" directory under your home directory. To avoid this,
    run the script as the root user, with "--sudo" option or with
    "--local-lib" option.

  cpanminus can't install the module XYZ. Is it a bug?
    It is more likely a problem with the distribution itself. cpanminus
    doesn't support or is known to have issues with distributions like as
    follows:

    *   Tests that require input from STDIN.

    *   Build.PL or Makefile.PL that prompts for input when
        "PERL_MM_USE_DEFAULT" is enabled.

    *   Modules that have invalid numeric values as VERSION (such as "1.1a")

    These failures can be reported back to the author of the module so that
    they can fix it accordingly, rather than me.

  Does cpanm support the feature XYZ of CPAN and CPANPLUS?
    Most likely not. Here are the things that cpanm doesn't do by itself.
    And it's a feature - you got that from the name *minus*, right?

    If you need these features, use CPAN, CPANPLUS or the standalone tools
    that are mentioned.

    *   CPAN testers reporting. See App::cpanminus::reporter

    *   Building RPM packages from CPAN modules

    *   Listing the outdated modules that needs upgrading. See
        App::cpanoutdated

    *   Showing the changes of the modules you're about to upgrade. See
        cpan-listchanges

    *   Patching CPAN modules with distroprefs.

    See cpanm or "cpanm -h" to see what cpanminus *can* do :)

COPYRIGHT
    Copyright 2010- Tatsuhiko Miyagawa

    The standalone executable contains the following modules embedded.

    CPAN::DistnameInfo Copyright 2003 Graham Barr
    local::lib Copyright 2007-2009 Matt S Trout
    HTTP::Tiny Copyright 2011 Christian Hansen
    Module::Metadata Copyright 2001-2006 Ken Williams. 2010 Matt S Trout
    version Copyright 2004-2010 John Peacock
    JSON::PP Copyright 2007-2011 by Makamaka Hannyaharamitu
    CPAN::Meta, CPAN::Meta::Requirements Copyright (c) 2010 by David Golden
    and Ricardo Signes
    CPAN::Meta::YAML Copyright 2010 Adam Kennedy
    File::pushd Copyright 2012 David Golden

LICENSE
    This software is licensed under the same terms as Perl.

CREDITS
  CONTRIBUTORS
    Patches and code improvements were contributed by:

    Goro Fuji, Kazuhiro Osawa, Tokuhiro Matsuno, Kenichi Ishigaki, Ian
    Wells, Pedro Melo, Masayoshi Sekimura, Matt S Trout (mst), squeeky,
    horus and Ingy dot Net.

  ACKNOWLEDGEMENTS
    Bug reports, suggestions and feedbacks were sent by, or general
    acknowledgement goes to:

    Jesse Vincent, David Golden, Andreas Koenig, Jos Boumans, Chris
    Williams, Adam Kennedy, Audrey Tang, J. Shirley, Chris Prather, Jesse
    Luehrs, Marcus Ramberg, Shawn M Moore, chocolateboy, Chirs Nehren,
    Jonathan Rockway, Leon Brocard, Simon Elliott, Ricardo Signes, AEvar
    Arnfjord Bjarmason, Eric Wilhelm, Florian Ragwitz and xaicron.

COMMUNITY
    <http://github.com/miyagawa/cpanminus> - source code repository, issue
    tracker
    <irc://irc.perl.org/#toolchain> - discussions about Perl toolchain. I'm
    there.

NO WARRANTY
    This software is provided "as-is," without any express or implied
    warranty. In no event shall the author be held liable for any damages
    arising from the use of the software.

SEE ALSO
    CPAN CPANPLUS pip



Changes
See http://github.com/miyagawa/cpanminus/ for the latest development.

1.7023  2015-01-05 07:59:52 JST
   [Improvements]
      - Bump Parse::PMFile to 0.34
      - Bump CPAN::Meta::Requirements to 2.131
      - Bump HTTP::Tiny to 0.53
      - Bump Module::Metadata to 1.000025
      - Bump local::lib to 2.000015

1.7022  2014-12-12 16:41:47 PST
   [Improvements]
      - Enabled unsafe option to parse version with PMFile to avoid segfaults and weird errors

1.7021  2014-12-11 21:42:36 PST
   [Improvements]
      - Bump Parse::PMFile to 0.32
      - Bump HTTP::Tiny to 0.052

1.7020  2014-12-08 17:53:18 PST
   [Improvements]
      - Bump Parse::PMFile to 0.31

1.7019  2014-12-04 12:51:41 PST
   [Improvements]
      - Bump Parse::PMFile to 0.30

1.7018  2014-11-25 14:08:41 PST
   [Maintenance]
      - bump $VERSION correctly in cpanm executable

1.7017  2014-11-25 14:00:40 PST
   [Improvements]
      - Bump fatpacked CPAN::Meta and HTTP::Tiny
      - Added documentation for -M and --from in man page

1.7016  2014-11-16 11:46:32 PST
   [New Features]
      - Added -M and --from to set --mirror and --mirror-only at the same time while
        overriding previously set mirrors (sri) #175, #417

   [Improvements]
      - cpanmin.us is now available under HTTPS (marcus)

1.7015  2014-11-14 13:14:07 PST
   [Improvements]
      - Bump Parse::PMFile to 0.29
      - Bump CPAN::Meta::Requirements to 2.129

1.7014  2014-10-07 20:52:58 PDT
   [Improvements]
      - Bump Parse::PMFile to 0.28

1.7013  2014-10-06 23:52:14 PDT
   [Improvements]
      - Update bundled JSON::PP to fix issues with JSON parsing on older perl 5.8

1.7012  2014-09-26 19:28:25 PDT
   [Improvements]
      - Upgrade CPAN::Meta and HTTP::Tiny in fatpack
      - Skip @INC on dev directory when running from git checkout when to search
        for an installed module

1.7011  2014-09-22 15:08:22 JST
      - Same as 1.7010. Make it a non-trial release.

1.7010  2014-09-17 18:27:53 JST
   [Improvements]
      - Use Parse::PMFile rather than our own fork (charsbar) #409

1.7009  2014-09-10 01:43:45 PDT
   [Bug Fixes]
      - Fix fatpacked executable to include Parse::CPAN::Meta

1.7008  2014-09-10 01:19:04 PDT
   [Improvements]
      - Show **** instead of *password* when masking passwords
      - Internal code refactoring

1.7007  2014-09-09 09:25:46 PDT
   [Improvements]
      - Make sure MetaCPAN search looks up the dist by right author #405
      - Ignore Module::Build artifacts in _build when indexing modules metadata

1.7006  2014-09-05 15:44:15 PDT
   [Bug Fixes]
      - Work around MetaCPAN bugs in numifying dev release versions (#367, #377)
      - Bump bootstrapping MakeMaker version recent enough to support MYMETA

   [Incompatible Changes]
      - Removed undocumented --metacpan option that was only for testing purposes

1.7005  2014-09-05 05:04:06 PDT
   [Bug Fixes]
      - Do not index common directories such as t, xt or inc in install.json
      - Support tarball extraction on HP-UX 11 (Brian Wightman)

   [Improvements]
      - Update Parse::PMFile to 0.19 equivalent (charsbar)
      - Fixed various documentation (syohex, Graham Ollis)
      - Mask passwords when failing to download a module (andyjones)
      - Don't find directories when searching for executables (haarg)
      - Only assume Module::Build configure prereq when there's no config prereq (haarg)
      - Upgrade fatpacked modules to the latest

1.7004  2014-04-27 18:22:11 CEST
   [Bug Fixes]
      - Downgrade Parse::CPAN::Meta as well #373

1.7003  2014-04-27 17:11:14 CEST
   [Bug Fixes]
      - Downgrade CPAN::Meta (moznion) #372

1.7002  2014-04-27 07:45:53 CEST
   [Bug Fixes]
      - Bump fatpacked local::lib #364
      - Fix a bug where build.log files are not properly renamed on the first run on Win32 (haarg) #368

1.7001  2013-09-08 13:11:16 PDT
   [Bug Fixes]
      - Fix bootstrap with perl 5.8 and MakeMaker/Module::Build (haarg) #311

1.7000  2013-09-04 15:01:34 PDT
      - Same as 1.6943. Some of the major changes between 1.61 are:
      - Add --uninstall/-U command
      - Support installing recommends/suggests/develop dependencies
      - Support features selection with --with-feature, --without-feature etc.
      - Support passing arguments to configure, build, test and install with --configure-args etc.
      - Add experimental --cpanfile option
      - Add experimental --pp option
      - Now much safer to run multiple instances of cpanm
      - Improved local::lib support
      - Improved MetaCPAN query
      - Improved version extraction for install metadata (for Carton)
      - Completely eliminated dependency on search.cpan.org

1.6943  2013-09-03 16:39:27 PDT
   [Bug Fixes]
      - Localize PERL_MB_OPT for --pureperl-only since it leaks out in build/test/install process
        (Thanks to doyster and leont)

1.6942  2013-08-27 11:10:40 PDT
   [Bug Fixes]
      - Fixed the date sort for developer releases with MetaCPAN #308

1.6941  2013-08-20 11:32:09 PDT
   [Incompatible Changes]
      - -v/--verbose doesn't imply --interactive anymore. It was added as a backward compatibility
        when --interactive was added, but i think it now does more harm than good. [RT#70041]

1.6940  2013-08-08 12:35:36 PDT
   [Bug Fixes]
      - Support extracting packages from dists without MANIFEST
        https://github.com/miyagawa/carton/issues/126

1.6939  2013-08-06 02:55:08 PDT
   [Bug Fixes]
      - Fix extracting version from Parse::RecDescent (lestrrat)

1.6938  2013-08-05 23:11:16 PDT
   [Bug Fixes]
      - Fix extracting version from modules with qv(). #299 (lestrrat)

1.6937  2013-08-05 18:54:31 PDT
   [Bug Fixes]
      - Fix a bug where version is not correctly scanned for VERSIONs using version.pm (lestrrat)

   [Improvements]
      - Gives a warning when local::lib directory contains spaces in its name, since it will
        likely cause issues with local::lib and MakeMaker #298

Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.029.
use strict;
use warnings;

use 5.008000;

use ExtUtils::MakeMaker;



my %WriteMakefileArgs = (
  "ABSTRACT" => "get, unpack, build and install modules from CPAN",
  "AUTHOR" => "- Tatsuhiko Miyagawa",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "App-cpanminus",
  "EXE_FILES" => [
    "bin/cpanm"
  ],
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.008000",
  "NAME" => "App::cpanminus",
  "PREREQ_PM" => {
    "ExtUtils::Install" => "1.46",
    "ExtUtils::MakeMaker" => "6.58",
    "Module::Build" => "0.38"
  },
  "TEST_REQUIRES" => {
    "Test::More" => 0
  },
  "VERSION" => "1.7023",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "ExtUtils::Install" => "1.46",
  "ExtUtils::MakeMaker" => "6.58",
  "Module::Build" => "0.38",
  "Test::More" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);





lib/App/cpanminus.pm
package App::cpanminus;
our $VERSION = "1.7006";

=encoding utf8

=head1 NAME

App::cpanminus - get, unpack, build and install modules from CPAN

=head1 SYNOPSIS

    cpanm Module

Run C<cpanm -h> or C<perldoc cpanm> for more options.

=head1 DESCRIPTION

cpanminus is a script to get, unpack, build and install modules from
CPAN and does nothing else.

It's dependency free (can bootstrap itself), requires zero
configuration, and stands alone. When running, it requires only 10MB
of RAM.

=head1 INSTALLATION

There are several ways to install cpanminus to your system.

=head2 Package management system

There are Debian packages, RPMs, FreeBSD ports, and packages for other
operation systems available. If you want to use the package management system,
search for cpanminus and use the appropriate command to install. This makes it
easy to install C<cpanm> to your system without thinking about where to
install, and later upgrade.

=head2 Installing to system perl

You can also use the latest cpanminus to install cpanminus itself:

    curl -L http://cpanmin.us | perl - --sudo App::cpanminus

This will install C<cpanm> to your bin directory like
C</usr/local/bin> (unless you configured C<INSTALL_BASE> with
L<local::lib>), so you probably need the C<--sudo> option.

=head2 Installing to local perl (perlbrew)

If you have perl in your home directory, which is the case if you use
tools like L<perlbrew>, you don't need the C<--sudo> option, since
you're most likely to have a write permission to the perl's library
path. You can just do:

    curl -L http://cpanmin.us | perl - App::cpanminus

to install the C<cpanm> executable to the perl's bin path, like
C<~/perl5/perlbrew/bin/cpanm>.

=head2 Downloading the standalone executable

You can also copy the standalone executable to whatever location you'd like.

    cd ~/bin
    curl -LO http://xrl.us/cpanm
    chmod +x cpanm
    # edit shebang if you don't have /usr/bin/env

This just works, but be sure to grab the new version manually when you
upgrade because C<--self-upgrade> might not work for this.

=head1 DEPENDENCIES

perl 5.8 or later.

=over 4

=item *

'tar' executable (bsdtar or GNU tar version 1.22 are recommended) or Archive::Tar to unpack files.

=item *

C compiler, if you want to build XS modules.

=item *

make

=item *

Module::Build (core in 5.10)

=back

=head1 QUESTIONS

=head2 Another CPAN installer?

OK, the first motivation was this: the CPAN shell runs out of memory (or swaps
heavily and gets really slow) on Slicehost/linode's most affordable plan with
only 256MB RAM. Should I pay more to install perl modules from CPAN? I don't
think so.

=head2 But why a new client?

First of all, let me be clear that CPAN and CPANPLUS are great tools
I've used for I<literally> years (you know how many modules I have on
CPAN, right?). I really respect their efforts of maintaining the most
important tools in the CPAN toolchain ecosystem.

However, for less experienced users (mostly from outside the Perl community),
or even really experienced Perl developers who know how to shoot themselves in
their feet, setting up the CPAN toolchain often feels like yak shaving,
especially when all they want to do is just install some modules and start
writing code.

=head2 Zero-conf? How does this module get/parse/update the CPAN index?

It queries the CPAN Meta DB site at L<http://cpanmetadb.plackperl.org/>.
The site is updated at least every hour to reflect the latest changes
from fast syncing mirrors. The script then also falls back to query the
module at L<http://metacpan.org/> using its wonderful API.

Upon calling these API hosts, cpanm (1.6004 or later) will send the
local perl versions to the server in User-Agent string by default. You
can turn it off with C<--no-report-perl-version> option. Read more
about the option with L<cpanm>, and read more about the privacy policy
about this data collection at L<http://cpanmetadb.plackperl.org/#privacy>

Fetched files are unpacked in C<~/.cpanm> and automatically cleaned up
periodically.  You can configure the location of this with the
C<PERL_CPANM_HOME> environment variable.

=head2 Where does this install modules to? Do I need root access?

It installs to wherever ExtUtils::MakeMaker and Module::Build are
configured to (via C<PERL_MM_OPT> and C<PERL_MB_OPT>). So if you're
using local::lib, then it installs to your local perl5
directory. Otherwise it installs to the site_perl directory that
belongs to your perl.

cpanminus at a boot time checks whether you have configured
local::lib, or have the permission to install modules to the site_perl
directory.  If neither, it automatically sets up local::lib compatible
installation path in a C<perl5> directory under your home
directory. To avoid this, run the script as the root user, with
C<--sudo> option or with C<--local-lib> option.

=head2 cpanminus can't install the module XYZ. Is it a bug?

It is more likely a problem with the distribution itself. cpanminus
doesn't support or is known to have issues with distributions like as
follows:

=over 4

=item *

Tests that require input from STDIN.

=item *

Tests that might fail when C<AUTOMATED_TESTING> is enabled.

=item *

Modules that have invalid numeric values as VERSION (such as C<1.1a>)

=back

These failures can be reported back to the author of the module so
that they can fix it accordingly, rather than me.

=head2 Does cpanm support the feature XYZ of L<CPAN> and L<CPANPLUS>?

Most likely not. Here are the things that cpanm doesn't do by
itself. And it's a feature - you got that from the name I<minus>,
right?

If you need these features, use L<CPAN>, L<CPANPLUS> or the standalone
tools that are mentioned.

=over 4

=item *

CPAN testers reporting. See L<App::cpanminus::reporter>

=item *

Building RPM packages from CPAN modules

=item *

Listing the outdated modules that needs upgrading. See L<App::cpanoutdated>

=item *

Showing the changes of the modules you're about to upgrade. See L<cpan-listchanges>

=item *

Patching CPAN modules with distroprefs.

=back

See L<cpanm> or C<cpanm -h> to see what cpanminus I<can> do :)

=head1 COPYRIGHT

Copyright 2010- Tatsuhiko Miyagawa

The standalone executable contains the following modules embedded.

=over 4

=item L<CPAN::DistnameInfo> Copyright 2003 Graham Barr

=item L<Parse::CPAN::Meta> Copyright 2006-2009 Adam Kennedy

=item L<local::lib> Copyright 2007-2009 Matt S Trout

=item L<HTTP::Tiny> Copyright 2011 Christian Hansen

=item L<Module::Metadata> Copyright 2001-2006 Ken Williams. 2010 Matt S Trout

=item L<version> Copyright 2004-2010 John Peacock

=item L<JSON::PP> Copyright 2007-2011 by Makamaka Hannyaharamitu

=item L<CPAN::Meta>, L<CPAN::Meta::Requirements> Copyright (c) 2010 by David Golden and Ricardo Signes

=item L<CPAN::Meta::YAML> Copyright 2010 Adam Kennedy

=item L<File::pushd> Copyright 2012 David Golden

=back

=head1 LICENSE

This software is licensed under the same terms as Perl.

=head1 CREDITS

=head2 CONTRIBUTORS

Patches and code improvements were contributed by:

Goro Fuji, Kazuhiro Osawa, Tokuhiro Matsuno, Kenichi Ishigaki, Ian
Wells, Pedro Melo, Masayoshi Sekimura, Matt S Trout (mst), squeeky,
horus and Ingy dot Net.

=head2 ACKNOWLEDGEMENTS

Bug reports, suggestions and feedbacks were sent by, or general
acknowledgement goes to:

Jesse Vincent, David Golden, Andreas Koenig, Jos Boumans, Chris
Williams, Adam Kennedy, Audrey Tang, J. Shirley, Chris Prather, Jesse
Luehrs, Marcus Ramberg, Shawn M Moore, chocolateboy, Chirs Nehren,
Jonathan Rockway, Leon Brocard, Simon Elliott, Ricardo Signes, AEvar
Arnfjord Bjarmason, Eric Wilhelm, Florian Ragwitz and xaicron.

=head1 COMMUNITY

=over 4

=item L<http://github.com/miyagawa/cpanminus> - source code repository, issue tracker

=item L<irc://irc.perl.org/#toolchain> - discussions about Perl toolchain. I'm there.

=back

=head1 NO WARRANTY

This software is provided "as-is," without any express or implied
warranty. In no event shall the author be held liable for any damages
arising from the use of the software.

=head1 SEE ALSO

L<CPAN> L<CPANPLUS> L<pip>

=cut

1;
